Mobs with `CanEscapeInventoryComponent` now have a grace period after escaping (or being dropped while escaping)

RadsammyT opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
I made mobs that are small enough to be in a players hands have a small period of time in which they cannot be picked up which takes effect if they were dropped after escaping or being dropped while escaping.
## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
This basically prevents players from dropping a mouse and then quickly picking them up again in order to effectively cancel the mouse's escape doAfter.
Also prevents AI Monkeys from trapping you in an infinite escape loop.

Also because this happened to me recently (as of making this PR) and I essentially made this PR out of spite.
## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
CanEscapeInventoryComponent and EscapeInventorySystem have been moved from Server to Shared, and made predictable.

In the component, added a `PenaltyTimer` TimeSpan.
In the system, Added a `GettingPickedUpEvent` Local Event Subscription that checks if `PenaltyTimer` is passed by `curTime`, changed `OnDropped()` so that it sets the `PenaltyTimer` when dropped WHILE a doAfter is in progress, changed `DoEscape()` to apply the grace period on success (because AI MONKEYS ARE A PAIN IN THE ASS)
## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.
-->
(penalty time is way larger than it should be for demoing purposes but its still gonna be 1 second)

https://github.com/user-attachments/assets/f74e261e-0f0c-4f24-b103-f6e099188c20


## Requirements
<!-- 
Due to influx of PR's we require to ensure that PR's are following the correct guidelines.

Please take a moment to read these if its your first time.

Check the boxes below to confirm that you have in fact seen these (put an X in the brackets, like [X]):
-->
- [X] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**

:cl:
- tweak: Mice, Mothroaches, etc., can no longer be picked up for a brief moment after escaping from your inventory or dropping them during.
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up. Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

RadsammyT committed:
initial working version.

my approach of removing and adding ItemComponent is gonna be so fucked

RadsammyT committed:
formatting

RadsammyT committed:
change PenaltyTimer to 1

RadsammyT committed:
do NOT hardcode 1f on assignment

worst mistake of my life

RadsammyT committed:
big :clap: review :clap:

use TimeSpan instead of deaccumulating frametime
make BasePenaltyTime a DataField

RadsammyT committed:
fuck it

by changing penaltyTimer to init with zero mobs with this component will
init unable to be picked up. but Integration tests fail so fuck it

RadsammyT committed:
REEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE

slarticodefast commented on Content.Server/Resist/CanEscapeInventoryComponent.cs:
```diff
@@ -11,6 +12,20 @@ public sealed partial class CanEscapeInventoryComponent : Component
+    /// <summary>
+    ///     Initial amount of time when you cannot be picked up when dropped while escaping
+    /// </summary>
+    [DataField("basePenaltyTime")]
```
```suggestion
    [DataField]
```
DataField names get automatically converted to camelCase nowadays. Old files still do this manually, but we don't need it anymore. You can do the same with the other DataField above.

slarticodefast commented on Content.Server/Resist/CanEscapeInventoryComponent.cs:
```diff
@@ -11,6 +12,20 @@ public sealed partial class CanEscapeInventoryComponent : Component
     [DataField("baseResistTime")]
```
```suggestion
    [DataField]
```

slarticodefast commented on Content.Server/Resist/EscapeInventorySystem.cs:
```diff
@@ -29,6 +34,20 @@ public override void Initialize()
+            if ( _timing.CurTime < comp.PenaltyTimer)
+                continue;
+            comp.PenaltyTimer = TimeSpan.MaxValue;
+            EnsureComp<ItemComponent>(uid);
```
You should not be removing the ItemComponent and add it later again. If you do that all the information stored inside it gets deleted. Instead you should subscribe to PickupAttemptEvent and cancel it if the penalty time is not over yet.

RadsammyT replied:
Gotcha. But `PickupAttemptEvent` is for the mob picking up the entity in question, not the entity with the EscapeInventory comp itself. Is `GettingPickedUpAttemptEvent` better?

slarticodefast replied:
Good catch. Yeah, that one sounds right

RadsammyT replied:
of course the downside with that (much better) approach is that this isnt predicted on the client but its also REALLY funny to see a lizard get japed trying to kidnap a mouse

RadsammyT replied:
should be resolved.

RadsammyT committed:
meme :clap: rememe :clap:

BaseResistTime and BasePenaltyTime use default DataField arguments
instead of hardcoded camelCase labels
removed ItemComponent fuckery to use GettingPickedUpAttemptEvent instead

RadsammyT committed:
formatting

slarticodefast commented:
Tested it and it works, but looks a little weird due to the mispredictions. I think this would work better better if you move the system and component to shared so it gets predicted properly and make the component networked as well (simply add the NetworkedComponent attribute for that). I think it should be possible since all the required events and systems are in Shared as well. The popups will have to be changed to use PopupPredicted in SharedPopupSystem for that. If you need any help, feel free to ping me on Discord (Slartibartfast). Otherwise a very solid first PR!

RadsammyT committed:
make predict

RadsammyT committed:
forgot to use PopupPredicted

slarticodefast commented:
That test fail happens randomly sometimes and is unrelated. I will rerun it for you.

RadsammyT commented:
> The popups will have to be changed to use PopupPredicted in SharedPopupSystem for that.

tried doing that (and i didnt test it out before I commited smh), but PopupPredicted also shows its message to everyone else in PVS range, which is prolly what I dont want.
keeping it to just SharedPopupSystem.PopupEntity(Msg, Ent, Recipient) demonstrates the behavior I want (respective messages shown to just their entities) but it dupes the popup.

RadsammyT committed:
keep PopupEntity

i never realized popups can also be painful

RadsammyT committed:
i forgor :skull:

beck-thompson commented:
Try PopupClient? I think that should just show it to the one client you want (I could be wrong)!

slarticodefast commented:
> Try PopupClient? I think that should just show it to the one client you want (I could be wrong)!

I tried that out as well, but the popup then does not show up for the player holding the mouse when it is attempting to escape using this:
`_popupSystem.PopupClient(Loc.GetString("escape-inventory-component-start-resisting-target"), container, container);`

Also there seems to be a desync. The doafter is started on both clients, but `MoveInputEvent` is never raised on the client holding the mouse when it escapes, causing the doafter not to be deleted. On the second mouse client and the server this runs as intended. I will do more testing on this, but maybe someone else knows the reason.

Golinth commented:
Mice will no longer be tormented by me, and will simply just die. Good luck, food.

RadsammyT commented:
PR has been one month stale now, but even with some consulting with Keron who made doAfters predicted weeks ago (starting [here](https://discord.com/channels/310555209753690112/310555209753690112/1278854249467613266) on #development channel at SS14 discord not much is progressing in terms of fixing the issue blocking this PR from being merged.

to recap, taking from the discord: there is an issue with letting the mouse escape, where `EscapeInventoryEvent` is NOT raised on the container entity causing a desync, resulting in the container entity thinking that the mouse can be picked up again immediately after escape (fact check: it wont). this does not happen when dropping the mouse while escaping.

thanks to debugging from @slarticodefast, quoting from them [here](https://discord.com/channels/310555209753690112/310555209753690112/1278867557276848268):
 > Yeah, I did all the breakpointing stuff. TryStartDoAfter returns true on both clients, we only never enter OnEscape on one of them. So it has to be somewhere in the doafter itself I think

if it comes down to it I may have to do the most masochistic thing of debugging on the desyncing client to see where in the doAfter code its failing at myself, but thats gonna take at most a week of headbanging for me.
